apiserver/internal/handlers/resource/resource.go

```diff
@@ -0,0 +1,417 @@
+			reader.Close()
+		}
+	}()
+
+	err = resourceService.StoreResourceAndIncrementCharmModifiedVersion(
+		req.Context(),
+		resource.StoreResourceArgs{
+			ResourceUUID:    uploaded.UUID,
+			Reader:          reader,
+			RetrievedBy:     username,
+			RetrievedByType: resource.User,
+			Size:            uploaded.Size,
+			Fingerprint:     uploaded.Fingerprint,
+			Origin:          charmresource.OriginUpload,
+			Revision:        -1,
+		},
+	)
+	if err != nil {
+		return nil, errors.Errorf("storing resource %s of application %s: %w", uploaded.Resource.Name, uploaded.Application, err)
+	}
+
+	res, err := resourceService.GetResource(req.Context(), uploaded.UUID)
+	if err != nil {
+		h.logger.Errorf("getting uploaded resource details: %w", err)
+		return &params.UploadResult{}, nil
+	}
+
+	return &params.UploadResult{
+		Resource: domainResource2API(res),
+	}, nil
+}
+
+// DomainResource2API converts a [domainresource.Resource] into
+// a [params.Resource] struct.
+func domainResource2API(res resource.Resource) params.Resource {
+	return params.Resource{
+		CharmResource: api.CharmResource2API(res.Resource),
+		ID:            res.UUID.String(),
+		ApplicationID: res.ApplicationID.String(),
+		Username:      res.RetrievedBy,
+		Timestamp:     res.Timestamp,
+	}
+}
+
+// uploadedResource holds both the information about an uploaded
+// resource and the reader containing its data.
+type uploadedResource struct {
+	// UUID is the resource UUID.
+	UUID coreresource.UUID
+
+	// Application is the Name of the application associated with the resource.
+	Application string
+
+	// Resource is the information about the resource.
+	Resource charmresource.Resource
+
+	// Size is the size of the resource blob.
+	Size int64
+
+	// Fingerprint is the hash of the resource blob.
+	Fingerprint charmresource.Fingerprint
+}
+
+// readResource extracts the relevant info from the request.
+func (h *ResourceHandler) readResource(resourceService ResourceService, req *http.Request) (io.ReadCloser, *uploadedResource, error) {
+	defer func() {
+		if req.Body != nil {
+			req.Body.Close()
+		}
+	}()
+	uReq, err := extractUploadRequest(req)
+	if err != nil {
+		return nil, nil, errors.Capture(err)
+	}
+
+	uuid, err := resourceService.GetResourceUUIDByApplicationAndResourceName(req.Context(), uReq.Application, uReq.Name)
+	if errors.Is(err, resourceerrors.ResourceNotFound) {
+		return nil, nil, jujuerrors.NotFoundf("resource %s of application %s", uReq.Name, uReq.Application)
+	} else if err != nil {
+		return nil, nil, errors.Errorf("getting resource uuid: %w", err)
+	}
+
+	res, err := resourceService.GetResource(req.Context(), uuid)
+	if err != nil {
+		return nil, nil, errors.Errorf("getting resource details: %w", err)
+	}
+
+	switch res.Type {
+	case charmresource.TypeFile:
+		ext := path.Ext(res.Path)
+		if path.Ext(uReq.Filename) != ext {
+			return nil, nil, errors.Errorf("incorrect extension on resource upload %q, expected %q", uReq.Filename, ext)
+		}
+	}
```

Comment:
This feels like it should be a service function. It might be worth locating it somewhere else.

Reply:
Agreed that it should be somewhere else, but I'm struggling to think of where. To be honest, it seems a bit weird that we even care about this, this is the only thing the filename is used for on the server, it seems like something that should be done client-side if anywhere.